pkg/components/discover/matcher.go

```diff
@@ -104,51 +104,80 @@ func (m *Matcher) filter(events []Event[ProcessAttrs]) []Event[ProcessMatch] {
 	return matches
 }
 
+func (m *Matcher) alreadyMatched(pid PID) bool {
+	_, ok := m.ProcessHistory[pid]
+	return ok
+}
+
+func (m *Matcher) matchCriteria(obj ProcessAttrs, proc *services.ProcessInfo) *ProcessMatch {
+	criteria := make([]services.Selector, 0, len(m.Criteria))
+
+	for i := range m.Criteria {
+		if m.matchProcess(&obj, proc, m.Criteria[i]) && !m.isExcluded(&obj, proc) {
+			criteria = append(criteria, m.Criteria[i])
+		}
+	}
+
+	if len(criteria) > 0 {
+		m.Log.Debug("found process", "pid", proc.Pid, "comm", proc.ExePath, "metadata",
+			obj.metadata, "podLabels", obj.podLabels, "criteria", criteria)
+
+		return &ProcessMatch{Criteria: criteria, Process: proc}
+	}
+
+	return nil
+}
+
 func (m *Matcher) filterCreated(obj ProcessAttrs) (Event[ProcessMatch], bool) {
-	if _, ok := m.ProcessHistory[obj.pid]; ok {
-		// this was already matched and submitted for inspection. Ignoring!
+	if m.alreadyMatched(obj.pid) {
 		return Event[ProcessMatch]{}, false
 	}
+
 	proc, err := processInfo(obj)
 	if err != nil {
 		m.Log.Debug("can't get information for process", "pid", obj.pid, "error", err)
 		return Event[ProcessMatch]{}, false
 	}
-	for i := range m.Criteria {
-		if m.matchProcess(&obj, proc, m.Criteria[i]) && !m.isExcluded(&obj, proc) {
-			m.Log.Debug("found process", "pid", proc.Pid, "comm", proc.ExePath, "metadata", obj.metadata, "podLabels", obj.podLabels, "criteria", m.Criteria[i])
-			m.ProcessHistory[obj.pid] = proc
-			return Event[ProcessMatch]{
-				Type: EventCreated,
-				Obj:  ProcessMatch{Criteria: m.Criteria[i], Process: proc},
-			}, true
-		}
+
+	processMatch := m.matchCriteria(obj, proc)
+
+	if processMatch != nil {
+		m.ProcessHistory[obj.pid] = *processMatch
```

Comment:
Nit: I'd restrict the `processMatch` lifespan, to avoid being alive together with `procMatch` below:
```go
if processMatch := m.matchCriteria(obj, proc); processMatch != nil {
```